### PR TITLE
feat: form metabox API enhancement for new form themes panel

### LINKED ISSUE #4500: feat: form metabox API enhancement for new form themes panel
# Feature Request

## Developer Story
As a developer, I want to update form metabox API I favor for form theme so that developers can create form themes easily which will support most of the core add-ons.

## Visuals
image

image

## Things to consider
As per @DevinWalker Suggestion: (https://www.figma.com/file/jaVXDA4wjQD9OqTi0l9KQe/Form-Themes?node-id=1882%3A44&viewport=-259%2C1586%2C0.5)
- **NEW FORM THEME TAB:**
This screen introduces the donor to the new “Form Theme” options. Here they can activate a theme and move into customizing it. 
New: We are removing the “Form Display” option and adding “Form Theme” tab
New: We are removing the “Form Content” tab and adding this below the specific theme options where applicable

- **Backwards Compat:** 
All existing forms will continue to use the “Legacy” form theme
New: A user cannot publish a form without first activating a form theme. If they attempt to they will be prompted with a validation notice. “Please select a form theme(linked to tab) to publish this donation form”.
New: The green notice at the bottom is a placeholder for now but we will want to promote feedback/interaction in some way. The survey is my first idea which would link to a page on our site.
Style: When hovering over a theme the preview screen image will pan up and down - see Astra starter sites directory for example

## Acceptance Criteria
- [x] We can register and save theme settings

### PR DESCRIPTION
**Pull request by ravinderk**
## Description
This Pr will resolve #4500 

This PR will resolve the remaining point of https://github.com/impress-org/givewp/pull/4507
- [x] https://github.com/impress-org/givewp/pull/4507#discussion_r388449887
```
So is there a context wherein you want to have every theme configuration loaded into memory? Or is it more likely that you just want to load the current active 
```

- [x] https://github.com/impress-org/givewp/pull/4507#discussion_r388451131
```
Please break $args up into individual parameters and then have one property per parameter. Having a single $this->data class property tells you nothing about the shape of the class. Your getters should be as simple as return $this->id.
```

- [x] https://github.com/impress-org/givewp/pull/4507#discussion_r388452121
```
This is really the point at which you could load the theme. Just have the $themeID be the same (or something you can derive to) the file name, then load the theme as needed.
```

## How Has This Been Tested?
This PR tested by updating the dummy thee form theme settings.

## Types of changes

Bug fix (non-breaking change which fixes an issue)

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style.
- [x] My code follows has proper inline documentation.

### CODE REVIEW

File: composer.json [hunk 1]
```diff
@@ -33,6 +33,9 @@
     "autoload": {
         "psr-4": {
             "Give\\": "src/"
-        }
+        },
+        "files": [
+            "src/Helpers/Helpers.php"
```
Comment (JasonTheAdams) [inline composer.json]: You're kind of recreating the file autoloading in the `Helpers.php` file right now by not actually including any functions in it. Just forego the `Helpers.php` file and just list out the files you want auto-included here in the `composer.json` files array. That way all autoloading is clearly in one place.

File: give.php [hunk 2]
```diff
@@ -755,6 +749,24 @@ private function is_request( $type ) {
...
+		 * @return mixed
+		 */
+		function __get( $propertyName ) {
+			if ( 'themes' === $propertyName ) {
+				if ( ! $this->themes instanceof Themes ) {
+					$this->themes = new Themes();
+				}
+
+				return $this->themes;
+			}
```
Comment (JasonTheAdams) [inline give.php]: So is this temporary until you get the other PR merged? Ideally you wouldn't have to list out the properties one by one in here, but have an array property you're able to reference (i.e. `$this->aliases`). Also, use something like a `$singletonsCache` array property. Avoid dynamic properties as they're a good way to run into unintentional collisions. So if this is temporary something like this:
```php
			if ( 'themes' === $propertyName ) {
				if ( ! isset($this->singletonCache[Themes::class] ) {
					$this->singletonCache[Themes::class] = new Themes();
				}

				return $this->singletonsCache[Themes::class]
			}
```

File: src/Form/Theme.php [hunk 3]
```diff
@@ -9,8 +9,8 @@
 
 namespace Give\Form;
 
-use function Give\Form\Theme\getActiveID;
-use function Give\Form\Theme\get;
+use InvalidArgumentException;
+use function Give\Helpers\Form\Theme\get;
```
Comment (JasonTheAdams) [inline src/Form/Theme.php]: When importing a function with a name that generic I suggest adding an alias for clarity:
```php
use function Give\Helpers\Form\Theme\get as getTheme;
```

File: src/Form/Themes.php [hunk 4]
```diff
@@ -26,11 +26,39 @@ class Themes {
 
 
 	/**
-	 * Get Registered themes
+	 * Themes constructor.
+	 */
+	public function __construct() {
+		$this->load();
```
Comment (JasonTheAdams) [inline src/Form/Themes.php]: No side-effects in the constructor. The act of constructing an object should _only_ construct that object. If an object has some initialization, then make that a method to be called: `$themes->loadThemes()`, and call that after instantiation:
```php
$themes = new Themes();
$themes->loadThemes();
```
You can do that when `Give` instantiates it.

File: composer.json (hunk 1 above)
Comment (ravinderk) [reply composer.json]: fixed: https://github.com/impress-org/givewp/compare/f60b190e8778...045cd0436d8c

File: src/Form/Theme.php (hunk 3 above)
Comment (ravinderk) [reply src/Form/Theme.php]: fixed: https://github.com/impress-org/givewp/compare/045cd0436d8c...7c2a09061baa

File: src/Form/Themes.php (hunk 4 above)
Comment (ravinderk) [reply src/Form/Themes.php]: https://github.com/impress-org/givewp/compare/7c2a09061baa...c96a437a6320

File: give.php (hunk 2 above)
Comment (ravinderk) [reply give.php]: @JasonTheAdams This was the only property that why I setup like it but I will use an array to store singleton and yes it is a temporary solution.

Comment (ravinderk) [reply give.php]: fix: https://github.com/impress-org/givewp/compare/c96a437a6320...b6184d567284

**Comment by ravinderk** [comment]
@JasonTheAdams Added fix for your suggestions.

File: give.php [hunk 5]
```diff
@@ -755,6 +758,25 @@ private function is_request( $type ) {
...
 
+		/**
+		 * Handle property get request
+		 *
+		 * @param string $propertyName
+		 *
+		 * @since 2.7.0
+		 * @return mixed
+		 */
+		function __get( $propertyName ) {
```
Comment (JasonTheAdams) [inline give.php]: One last thing here, so yourself (and everyone else) a favor and add this to the docblock for the class:
```php
@property-read Themes $themes
```
That will inform all IDE's that the read-only property exists and what it returns.

File: src/Form/Themes.php [hunk 6]
```diff
@@ -26,11 +26,38 @@ class Themes {
 
 
 	/**
-	 * Get Registered themes
+	 * Themes constructor.
+	 */
+	public function __construct() {
+	}
```
Comment (JasonTheAdams) [inline src/Form/Themes.php]: Now you can remove this since you're no longer using it.

Comment (ravinderk) [reply src/Form/Themes.php]: https://github.com/impress-org/givewp/compare/b6184d567284...40544c3ae37d

File: give.php (hunk 5 above)
Comment (ravinderk) [reply give.php]: https://github.com/impress-org/givewp/compare/40544c3ae37d...1d3d0f8d03b3

**Comment by ravinderk** [comment]
@JasonTheAdams Issue resolved as per your suggestion.

**Pull request merged**